topdown/topdown_test.go

```diff
@@ -2637,12 +2638,65 @@ func TestTopDownWithKeyword(t *testing.T) {
 			rules: []string{`p = true { data.ex.loopback with input.foo as "x" with input.foo.bar as "y" }`},
 		},
 		{
-			note: "With invalidate",
+			note: "with invalidate",
 			exp:  `[2,3,4]`,
 			modules: []string{`package ex
 			input_eq = true { input.x = 1 }`},
 			rules: []string{`p[x] { data.a[_] = x; not data.ex.input_eq with input.x as x }`},
 		},
+		{
+			note:  "with invalidate input stack",
+			exp:   `["a", "b"]`,
+			input: `"b"`,
+			rules: []string{
+				`p = [x, y] { x = input with input as "a"; y = input }`,
+			},
+		},
+		{
+			note:  "with invalidate input stack iteration",
+			exp:   `[["a", "c"], ["b", "c"]]`,
+			input: `"c"`,
+			rules: []string{
+				`q[x] { input[_] = x }`,
+				`p[[x,y]] {
+					q[x] with input as ["a", "b"]
+					y = input
+				}`,
+			},
+		},
+		{
+			note:  "with invalidate virtual cache",
+			exp:   `["a", "b"]`,
+			input: "2",
+			rules: []string{
+				`q = "a" { input = x; x = 1 }`,
+				`q = "b" { input = x; x = 2 }`,
+				`p = [x, y] {
+					q = x with input as 1
+					q = y
+				}`},
+		},
+		{
+			note: "with invalidate data stack",
+			exp:  `["a", "b"]`,
+			rules: []string{
+				`q = "b" { true }`,
+				`p = [x ,y] {
+					q = x with q as "a"
+					q = y
+				}`,
+			},
+		},
+		{
+			note: "with invalidate data stack iteration",
```

Comment:
What is the `exp` result ?

Reply:
Nice catch! I've fixed this and improved the test helper to error if an unexpected expected value is received (in this case, the helper was receiving `nil` for the expected value which just fell through the switch statement).